### ISSUE TITLE
[API Proposal]: Simplify lookup of parsed TypeName in metadata

### ISSUE THREAD

**Issue by jkotas** [dotnet/runtime#101774]
### Background and motivation

The recently introduced System.Reflection.Metadata.TypeName is likely to be frequently combined with lookup of the type in metadata. The information required to lookup a type in ECMA335 metadata tables is unescaped (namespace, name) tuple. This information is not provided by the `TypeName` type today. In order to lookup the parser type name in metadata, the consumers have to implement unescaping and full type name splitting helper methods to convert the information provided `TypeName` this tuple. 

### API Proposal

```diff
  namespace System.Reflection.Metadata;

  public class TypeName
  {
      // Existing APIs. Return escaped names (to match reflection), no API to get Namespace only.
      public string Name { get; }
      public string FullName { get; }
      
      // Returns (unescaped namespace, unescaped name) tuple for simple types. Throws exception otherwise.
      // Namespace is an empty string for nested types.
+     public (string typeNamespace, string typeName) GetMetadataNamespaceAndName();

  }
```

### API Usage

```csharp
CustomType ResolveType(TypeName typeName)
{
    if (typeName.IsSimple)
    {
        if (typeName.IsNested)
        {
            (_, string nestedName) = typeName.GetMetadataNamespaceAndName();
            ... use nestedName to lookup the type in metadata ...
        }
        else
        {
            (string typeNamespace, string typeName) = typeName.GetMetadataNamespaceAndName();
            ... use typeNamespace and typeName to lookup the type in metadata ...
        }
    }

    ... handle other cases ...
}
```

### Alternative Designs

Introduce static helper methods in System.Reflection.Metadata for unescaping of a type name and splitting type name into a (namespace, name) tuple.

**Comment by jkotas** [comment]
I have filled this proposal based on my experience with implementing lookup of parsed TypeName in tools https://github.com/dotnet/runtime/pull/101767.

@adamsitnik Any thoughts about this API?

**Comment by jkotas** [comment]
Good point. .NET Framework and .NET Core never followed this rule in practice. The full type name is expected to be split in namespace + name using [this algorithm](https://github.com/dotnet/runtime/blob/ee93104aa0e7ac77ca7e4c6db9873356a9e3e622/src/coreclr/utilcode/namespaceutil.cpp#L92-L106). We may want to add a note to https://github.com/dotnet/runtime/blob/main/docs/design/specs/Ecma-335-Augments.md .

**Comment by adamsitnik** [comment]
>  Returns (unescaped namespace, unescaped name) tuple for simple types. 

Does it always return `Name` or `FullName`? Or does it depend on the nested vs non-nested?

> tuple

This particular library can be consumed by Full Framework apps and they often struggle to use `ValueTuple` (https://github.com/dotnet/BenchmarkDotNet/issues/895, https://github.com/dotnet/BenchmarkDotNet/issues/667)

> Throws exception otherwise.

I see why low level metadata parsers need that, but I find it too specific to particular use case.

What would we benefit from exposing such public API?

**Comment by jkotas** [comment]
> Does it always return Name or FullName? Or does it depend on the nested vs non-nested?

For non-nested types, it returns unescaped FullName split into name and namespace.
For nested types, it returns unescaped Name (nested types do not have namespace)

> This particular library can be consumed by Full Framework apps and they often struggle to use ValueTuple

I do not mind changing this to whatever else API review prefers; or expose the static helper methods instead.

> I see why low level metadata parsers need that, but I find it too specific to particular use case.

System.Reflection.Metadata is a library for low-level metadata readers. Every tool or library that reads metadata in its entirety using this library needs this functionality. Every shipping `TypeName` use in this repo needs it so far.

Also, I was looking at converting https://github.com/dotnet/runtime/blob/71abc26cf8f14f5d15665d7f580365ddfa296921/src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/Helpers.cs#L293-L330 to use TypeName. It would need to include the .cs files with the static helper methods in yet another place if I do that.

**Comment by adamsitnik** [comment]
> System.Reflection.Metadata is a library for low-level metadata readers. 

You have convinced me 👍 

FWIW the alternative design for me:

```diff
namespace System.Reflection.Metadata;

public class TypeName
{
     public string Name { get; }
     public string FullName { get; }
    
+    public string Namespace { get; }
+    public static string Unescape(string name);
}
```

**Comment by jkotas** [comment]
Update the proposal at the top. I think it is a better starting point for the API review discussion.

**Comment by steveharter** [comment]
Moving to v10 since we have reached the "feature complete" milestone; this is still on the API review backlog.

**Comment by bartonjs** [comment]
* System.Type.Namespace is a nullable string property, and seems to have distinction between `null` (Generic type parameters) and empty (global namespace types); it was asked if we need the same semantics here.  The answer was "no, this will always return empty".
* It was asked if we need a `static string Escape(string name)` for symmetry, and the answer was "no, call the CreateSimpleTypeName API and then read the appropriate name property from the result".

```c#
namespace System.Reflection.Metadata;

public partial class TypeName
{      
    public string Namespace { get; }

    public static string Unescape(string name);
}
```
